opalog/parser_test.go

```diff
@@ -79,39 +132,64 @@ func testParse1TermFail(t *testing.T, msg string, expr string) {
 	}
 }
 
-func TestScalarTerms(t *testing.T) {
-	testParse1Term(t, "null", "null", NewNull())
-	testParse1Term(t, "true", "true", GoTerm(true))
-	testParse1Term(t, "false", "false", GoTerm(false))
-	testParse1Term(t, "integer", "53", GoTerm(53))
-	testParse1Term(t, "integer2", "-53", GoTerm(-53))
-	testParse1Term(t, "float", "16.7", GoTerm(16.7))
-	testParse1Term(t, "float2", "-16.7", GoTerm(-16.7))
-	testParse1Term(t, "exponent", "6e7", GoTerm(6e7))
-	testParse1Term(t, "string", "\"a string\"", GoTerm("a string"))
-	testParse1Term(t, "string", "\"a string u6abc7def8abc0def with unicode\"",
-		GoTerm("a string u6abc7def8abc0def with unicode"))
-
-	testParse1TermFail(t, "hex", "6abc")
-	testParse1TermFail(t, "non-string", "'a string'")
-	testParse1TermFail(t, "non-bool", "True")
-	testParse1TermFail(t, "non-bool", "False")
-	testParse1TermFail(t, "non-number", "6zxy")
-	testParse1TermFail(t, "non-number2", "6d7")
+func newObjectTerm(o []*KeyValue) *Term {
+	set := NewKeyValueSet()
+	for _, v := range o {
+		set.Add(v)
+	}
+	return NewTerm(set, OBJECT, []byte(""), "", 0, 0)
 }
 
-func TestObjectTerms(t *testing.T) {
-	correct := GoTerm(map[string]int{"abc": 7, "def": 8})
-	testParse1Term(t, "simple dict", "{\"abc\": 7, \"def\": 8}", correct)
+func newArrayTerm(arr []*Term) *Term {
+	return NewTerm(arr, ARRAY, []byte(""), "", 0, 0)
 }
 
-// func TestVariables(t *testing.T) {
-// 	testParse(t, "variable", "\"a string\"")
-// }
+func reflectTerm(x interface{}) *Term {
+
+	if x == nil {
+		return NewTerm(nil, NULL, []byte(""), "", 0, 0)
+	}
+
+	if v, ok := x.(*Var); ok {
+		return NewTerm(v, VAR, []byte(""), "", 0, 0)
+	}
+
+	var val interface{}
+	var typ int
+	switch reflect.TypeOf(x).Kind() {
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
+		reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		val = float64(reflect.ValueOf(x).Int())
+		typ = NUMBER
+	case reflect.Float32, reflect.Float64:
+		val = float64(reflect.ValueOf(x).Float())
+		typ = NUMBER
+	case reflect.String:
+		val = x
+		typ = STRING
+	case reflect.Bool:
+		val = x
+		typ = BOOLEAN
+	case reflect.Map:
+		kvset := NewKeyValueSet()
+		xval := reflect.ValueOf(x)
+		for _, key := range xval.MapKeys() {
+			kvset.Add(NewKeyValue(reflectTerm(key.Interface()), reflectTerm(xval.MapIndex(key).Interface())))
+		}
+		val = kvset
+		typ = OBJECT
+	case reflect.Slice, reflect.Array:
+		xval := reflect.ValueOf(x)
+		length := xval.Len()
+		arr := make([]*Term, length)
+		for i := 0; i < length; i++ {
+			arr[i] = reflectTerm(xval.Index(i).Interface())
+		}
+		val = arr
+		typ = ARRAY
+	default:
+		panic(fmt.Sprintf("Unexpected type of term: %v", x))
+	}
 
```

Comment:
Feel free to delay this to another commit, but we're not testing combinations of embedded composites, such as [[1,foo,3], {"abc": bar}].